### PR TITLE
gh-116167: Allow disabling the GIL with `PYTHON_GIL=0` or `-X gil=0`

### PR DESCRIPTION
**Pull request by swtaarrs**
In free-threaded builds, running with `PYTHON_GIL=0` or `-X gil=0` will now disable the GIL. #116322 and #116329 track follow-up work to re-enable the GIL when loading an incompatible extension, and to disable the GIL by default, respectively.

In order to support re-enabling the GIL at runtime, all GIL-related data structures are initialized as usual, and disabling the GIL simply sets a flag at startup that causes `take_gil()` and `drop_gil()` to return early.

With `PYTHON_GIL=0` set, I spot-checked a few tests and small programs that don't use threads. They all seem to run fine, and very basic threaded programs work, sometimes. Trying to run the full test suite crashes pretty quickly, in `test_asyncio`.

- Issue: gh-116167

### CODE REVIEW

**Comment by swtaarrs** [comment]
!buildbot nogil

**Comment by swtaarrs** [comment]
!buildbot nogil

**Review by corona10 — COMMENTED** [review]
Good to add test case in this file too.
https://github.com/python/cpython/blob/72714c0266ce6d39c7c7fb63f617573b8f5a3cb2/Lib/test/test_cmd_line.py#L896

**Comment by corona10** [comment]
Ah ignore above comments, this PR only referencing the environment variable.

**Comment by corona10** [comment]
At not for this PR, since this feature is not officially announced, but in some moment, you should write about the environment variable at https://github.com/python/cpython/blob/72714c0266ce6d39c7c7fb63f617573b8f5a3cb2/Doc/using/cmdline.rst to get user feedback.

**Comment by swtaarrs** [comment]
!buildbot nogil

**Comment by swtaarrs** [comment]
> you should write about the environment variable ... to get user feedback

What would be a good time to do that? Maybe once the GIL is disabled by default, so the description doesn't have to change?

File: Python/initconfig.c [hunk 1]
```diff
@@ -1642,6 +1651,21 @@ config_read_env_vars(PyConfig *config)
...
+    const char *gil = config_get_env(config, "PYTHON_GIL");
+    if (gil != NULL) {
+#ifdef Py_GIL_DISABLED
+        if (strcmp(gil, "0") == 0) {
+            config->enable_gil = _PyConfig_GIL_DISABLE;
+        } else if (strcmp(gil, "1") == 0) {
+            config->enable_gil = _PyConfig_GIL_ENABLE;
+        } else {
+            return _PyStatus_ERR("PYTHON_GIL must be \"0\" or \"1\"");
+        }
```
Comment (colesbury) [inline Python/initconfig.c]: Format with `else if` and `else` on new lines (see [PEP 7](https://peps.python.org/pep-0007/#:~:text=Braces%20should%20be%20formatted%20as%20shown%3A))

**Comment by colesbury** [comment]
I would document the environment variable now, when it's added. For reference, we already document the [`--disable-gil`](https://docs.python.org/3.13/using/configure.html#cmdoption-disable-gil) configure flag.

I don't think it needs to be extensive, but should cover:
- Only available in free-threaded builds
- `0` forces the GIL to be disabled
- `1` forces the GIL to be enabled
- New in version 3.13

The behavior when it is not set is what's going to change in the future. We can document that later.

**Comment by swtaarrs** [comment]
After discussion in Discord, I'm also going to add a `-X` option to go with the environment variable.

File: Lib/test/_test_embed_set_config.py [hunk 2]
```diff
@@ -211,6 +212,14 @@ def test_flags(self):
         self.set_config(use_hash_seed=1, hash_seed=123)
         self.assertEqual(sys.flags.hash_randomization, 1)
 
+        if support.Py_GIL_DISABLED:
+            self.set_config(enable_gil=-1)
+            self.assertEqual(sys.flags.gil, None)
+            self.set_config(enable_gil=0)
+            self.assertEqual(sys.flags.gil, 0)
+        self.set_config(enable_gil=1)
```
Comment (ericsnowcurrently) [inline Lib/test/_test_embed_set_config.py]: FWIW, as a reader this implies to me that `PyConfig.enable_gil` is available even on builds without `--disable-gil`.  Seeing that it isn't available makes me wonder why this call isn't failing.  Is `self.set_config()` tolerant of unrecognized config fields?

**Review by ericsnowcurrently — APPROVED** [review]
LGTM

I've left one minor comment about an unclear test.  It really isn't a blocker, as the test is passing, but it would still be nice to have a response.

**Review by erlend-aasland — APPROVED** [review]

**Review by corona10 — APPROVED** [review]

File: Lib/test/_test_embed_set_config.py (hunk 2 above)
Comment (swtaarrs) [reply Lib/test/_test_embed_set_config.py]: > Is `self.set_config()` tolerant of unrecognized config fields?

Oops, yeah. It ends up calling [`_PyConfig_FromDict()`](https://github.com/python/cpython/blob/main/Python/initconfig.c#L1285), which looks up every key from `PYCONFIG_SPEC` in the given dict, never checking for superfluous keys.

I wasn't intentionally using this behavior; I just forgot that I had conditionally defined `PyConfig.enable_gil` while writing this line of the test. I can remove the `set_config()` call and add a comment.

**Review by colesbury — APPROVED** [review]

**Comment by alikud** [comment]
Writing in legendary merge request

**Comment by edpyt** [comment]
wow

**Comment by fractalvision** [comment]
Just leaving the comment for the history. Genuinely epic thread, what a time to be alive!

**Pull request merged**